File: scala/src/main/scala/folio/codinginterview/infrastructure/repository/AccountRepositoryImpl.scala
package folio.codinginterview.infrastructure.repository

import folio.codinginterview.application.repository.AccountRepository
import folio.codinginterview.domain.Account
import folio.codinginterview.domain.UserId
import scala.collection.concurrent.TrieMap
import scala.concurrent.Future

final class AccountRepositoryImpl extends AccountRepository {
  private val store: TrieMap[String, Account] = TrieMap.empty

  override def find(userId: UserId): Future[Option[Account]] =
    Future.successful(store.get(userId.value))

  override def upsert(userId: UserId, account: Account): Future[Unit] = {
    store.update(userId.value, account)
    Future.unit
  }

  override def exists(userId: UserId): Future[Boolean] =
    Future.successful(store.contains(userId.value))
}

File: php/src/CodingInterview/Infrastructure/Repository/AccountRepositoryImpl.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Infrastructure\Repository;

use Folio\CodingInterview\Application\Repository\AccountRepository;
use Folio\CodingInterview\Domain\Account;
use Folio\CodingInterview\Domain\UserId;

final class AccountRepositoryImpl implements AccountRepository
{
    /** @var array<string,Account> */
    private array $store = [];

    public function find(UserId $userId): ?Account
    {
        return $this->store[$userId->value] ?? null;
    }

    public function upsert(UserId $userId, Account $account): void
    {
        $this->store[$userId->value] = $account;
    }

    public function exists(UserId $userId): bool
    {
        return array_key_exists($userId->value, $this->store);
    }
}
